### PR TITLE
Crosswalk for Kubernetes

### PR DESCRIPTION
**Pull request by metral**
- Branch renamed to `crosswalk`
- Supersedes PR https://github.com/pulumi/docs/pull/1850
  - See previous PR relevant discussion and comments

---

[Live Docs Preview Link](http://content-bucket-268a8c6.s3-website-us-east-1.amazonaws.com/docs/guides/crosswalk/kubernetes/)
  - Note, the live preview is an active WIP that will change often during development. For an up-to-date render, run this docs PR locally.

### CODE REVIEW

**Comment by justinvp** [comment]
@metral, I haven't looked into this PR to see if you've already done this, but would it be worth ensuring we have aliases on new pages for all the old URLs at `/docs/guides/k8s-the-prod-way/*`?  I know we didn't publicize the old URLs in any top-level-way, but we did link to them from one or two places previously (kind of buried), so it'd be nice if visiting the old URLs didn't 404 but instead redirected to new locations (with similar or same content). Or, if there isn't a good 1:1 mapping, just add aliases to `content/docs/guides/crosswalk/kubernetes/_index.md` for all the old URLs.

**Comment by metral** [comment]
@justinvp 

>  but would it be worth ensuring we have aliases on new pages for all the old URLs at /docs/guides/k8s-the-prod-way/*?

IIRC I removed only a single link to it that we had in the kubernetes cloud provider page, but no other spots ref'd it that I could tell, and testing is passing with no broken URLs, so I don't believe it was widely linked to internally.

The guides were inspired by the original ktpw work, but it's spun off into its own thing that doesn't quite match up with the original work 1:1, so I'm not sure if alias would be all too relevant.

Thoughts @lukehoban?

**Comment by justinvp** [comment]
> IIRC I removed only a single link to it that we had in the kubernetes cloud provider page, but no other spots ref'd it that I could tell

Yeah, I'm pretty sure that was the one place. My point is: those pages were public and crawlable by Google, so it'd be nice if visiting the old URLs took you to the new guide (even if it's just the index), rather than 404ing.

If I search Google for "kubernetes the prod way" some of the old pages still show up:

Screen Shot 2019-10-21 at 1 17 07 PM

**Comment by lukehoban** [comment]
> Thoughts @lukehoban?

Agree with @justinvp - a handful of aliases to try to map the old pages roughly into the new content (even if imprecise) will help ensure we don't have dangling links from the rest of the internet.

**Comment by joeduffy** [comment]
Minor nit, I noticed the AWS | Azure | GCP  tabs don't have a hover pointer correctly, which makes them not look clickable. Can we add the proper CSS to do this? (The language/etc tabs do this.) Guessing @cnunciato can help if necessary.

**Comment by cnunciato** [comment]
Yep, just add `.mdl-tabs__tab.cloud-tab`  to this selector list:

https://github.com/pulumi/docs/blob/810fd33e6ee78db2c08e533046cfe676e0f64bdd/assets/sass/_chooser.scss#L116

File: assets/js/chooser.js [hunk 1]
```diff
@@ -138,4 +148,6 @@ $(function() {
         defaultOsChoice = "linux";
     }
     hideShowChoices("os", selectOs, defaultOsChoice);
+    hideShowChoices("cloud", selectCloud, "aws");
+    hideShowChoices("k8s-language", selectK8sLang, "yaml");
```
Comment (joeduffy) [inline assets/js/chooser.js]: Feels like we should default to the standard Pulumi model, and let folks choose YAML if they want to compare against "classic" -- or Kx if they want to see the nicer variant.

Comment (metral) [reply assets/js/chooser.js]: Done

**Comment by lblackstone** [comment]
@metral Probably want to add `pulumi/kubernetesx` to the diagram at docs/guides/crosswalk/kubernetes/#production-infrastructure-as-code

File: content/docs/guides/crosswalk/kubernetes/configure-defaults.md [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+$ kubectl apply -f privileged-rolebinding.yaml
+```
+{{% /md %}}
+</div>
+
+<div class="cloud-prologue-gcp"></div>
+<div class="mt">
+{{% md %}}
+
+```yaml
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/configure-defaults.md]: Add Pulumi-native example:
```ts
const privilegedCRB = new k8s.rbac.v1.ClusterRoleBinding("privileged", {
    metadata: { name: "allow-privileged-ingress-nginx" },
    roleRef: {
        apiGroup: "rbac.authorization.k8s.io",
        kind: "ClusterRole",
        name: "gce.privileged"
    },
    subjects: [
        {
            kind: "Group",
            name: "system:serviceaccounts:ingress-nginx",
            apiGroup: "rbac.authorization.k8s.io"
        }
    ]
});
```

File: content/docs/guides/crosswalk/kubernetes/configure-access-control.md [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+### Use the `devs` ServiceAccount
+
+In [Identity][crosswalk-identity] we demonstrate how to create typical roles for
+use in Kubernetes.
+
+We created an `devs` ServiceAccount for cluster administrators with root privileges, that
+will be tied into Kubernetes RBAC.
+
+Authenticate as the `devs` ServiceAccount from the [Identity
+stack][gcp-identity-stack].
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/configure-access-control.md]: Broken link

File: content/docs/guides/crosswalk/kubernetes/app-services.md [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+
+The full code for the general app services is on [GitHub][gh-repo-stack].
+[gh-repo-stack]: https://github.com/pulumi/kubernetes-the-prod-way/tree/crosswalk/general-app-services
+
+
+## Overview
+
+We'll explore how to setup:
+
+  * [Datstores](#datastores)
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/app-services.md]:
```suggestion
  * [Datastores](#datastores)
```

File: content/docs/guides/crosswalk/kubernetes/apps.md [hunk 5]
```diff
@@ -51,11 +50,12 @@ Check out how to:
 
   * [Build and Deploy a Container](#build-and-deploy-a-container)
   * [Deploy a Pod with a Sidecar](#deploy-a-pod-with-a-sidecar)
+  * [Deploy a Helm Chart](#deploy-a-helm-chart)
   * [Deploy Wordpress](#deploy-wordpress)
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/apps.md]: It looks like you might have removed this section, so this link can go, I think.

Comment (metral) [reply content/docs/guides/crosswalk/kubernetes/apps.md]: It's in there :)

Comment (lblackstone) [reply content/docs/guides/crosswalk/kubernetes/apps.md]: Ha, you're right. I couldn't find it last night, but that may have been a lack of sleep.

**Comment by lblackstone** [comment]
Looks like we still need a top level link at `/docs/guides/`
Screen Shot 2019-11-13 at 10 02 30 AM

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  * [Configure Access Control][crosswalk-configure-access]
+  * [Deploy Cluster Services][crosswalk-cluster-svcs]
+  * [Deploy App Services][crosswalk-app-svcs]
+  * [Deploy Apps][crosswalk-apps]
+  * [Update the Worker Nodes][crosswalk-update-worker-nodes]
+
+## Production Architecture for Teams
+
+Misconfigured infrastructure accounts are the source of a significant number of serious
+production outages in major systems. Many of these failure modes are preventable.
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
production outages. Many of these failure modes are preventable.
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+Pulumi tools allow engineering teams to share specifications for what their
+infrastructure should look like and allow teams to reliably provision and manage
+infrastructure. Changes to infrastructure can be audited as part of
+code review and they allow teams to detect drift.
+
+This architecture is meant to show how these tools can be used within a team
+to employ and understand:
+
+* **Security:** Who has access to what, and how are things ensured?
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
* **Security:** Who has access to what, and how is this policy enforced?
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+The identity stack typically contains:
+
+* Identities and roles for the team e.g. [AWS IAM][aws-iam], [GCP IAM][gcp-iam], [Azure AD][azure-ad].
+
+    For example, the database team typically gets only administrative permissions for the datastores, while an app team might only get cluster developer permissions.
+* Service Accounts for bots and CI/CD.
+
+    While IAM roles and Active Directory accounts describe identity of users,
+    service accounts grant an identity for workloads, e.g. Storage
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
    service accounts grant an identity for workloads, e.g., Storage
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+Provisioning shared, managed infrastructure is required to configure the
+cluster.
+
+At a minimum, this typically includes networking infrastructure,
+and can often include storage backends along with other cloud services such as
+VMs, registries, data pipelines, and data warehouses.
+
+### 3. Kubernetes Cluster
+
+Configuring and provision the Kubernetes cluster with the desired settings and defaults.
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
Configure and provision the Kubernetes cluster with the desired settings and defaults.
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+This also typically involves provisioning the Kubernetes cluster infrastructure
+with [API resources][k8s-api-resources] such as Namespaces, Roles , RoleBindings, and Quotas.
+
+Using a managed Kubernetes cluster on [EKS][eks], [GKE][gke], or [AKS][aks] is
+the easiest recommended way to get started with deploying a cluster.
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
the easiest way to deploy a cluster.
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+### 4. Cluster Services
+
+With a running, vanilla cluster you can install any Kubernetes cluster-scoped
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
With a vanilla cluster running, you can install any Kubernetes cluster-scoped
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+services that will be shared by a subset or all cluster users.
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
services that will be shared by some or all cluster users.
```

File: content/docs/guides/crosswalk/kubernetes/_index.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+At a minimum, services that should be installed include centralized cluster and app-based logging, and often
+include monitoring, policies, and service meshes.
+
+### 5. App Services
+
+Configure any Kubernetes app-scoped services that will be shared
+by a subset or all users with deployment permissions.
```
Comment (lblackstone) [inline content/docs/guides/crosswalk/kubernetes/_index.md]:
```suggestion
with users using deployment permissions.
```

**Comment by metral** [comment]
> Looks like we still need a top level link at /docs/guides/

Good catch. Done

**Pull request merged**